src/generator/generator_visitor.ts

```diff
@@ -339,28 +328,40 @@ class GeneratorVisitor extends Visitor {
         const value = this.popExpression();
 
         if (isIdentifier(val)) {
-            const id = this.variableMapping.get(val.name);
-
-            if (id === undefined) {
-                throw new Error('Assigned to unknown variable');
-            }
-
-            this.statements.push(this.module.set_local(id, value));
+            this.setLocal(val, value);
         } else if (isMemberExpression(val)) {
-            this.visit(val.object);
-            this.visit(val.property);
-
-            const index = this.popExpression();
-            const arrayPointer = this.popExpression();
-            const address = this.module.i32.add(arrayPointer, this.module.i32.mul(index, this.module.i32.const(4)));
-
-            // @ts-ignore because store() returns an expression
-            this.statements.push(this.module.i32.store(0, 4, address, value));
+            this.setArrayElement(val, value);
         } else {
             throw new Error('Assignment to non-identifier or member expression');
         }
     }
 
+    private setLocal(identifier: Identifier, value: Expression) {
+        this.statements.push(this.module.set_local(this.getVariableIndex(identifier.name), value));
+    }
+
+    private getArrayElement(node: MemberExpression) {
+        super.visitMemberExpression(node);
+
+        // The offset can not be used, because the memberaccess value can also be a mathematical term or a variable
+        this.expressions.push(this.module.i32.load(0, 4, this.getAdress()));
+    }
+
+    private getArrayLength(node: MemberExpression) {
```

Comment:
Can't you use the `getAdress` method here as well?

Reply:
Yes I could, if I push a constant `-1` onto the expression stack. The problem I have with such an implementation is that it would not improve readability. If the binaryen optimizer does not inline the multiplication of the two constant values we would even loose some performance.